Docs/BazelForiOSDevelopers.md

```diff
@@ -0,0 +1,370 @@
+documentation for all possible arguments. _Note: Unlike Starlark rules which
+are easily added on to Bazel, the `objc_library` is part of the internal java
+rules shipped with the Bazel
+binary._ 
+
+### Toolchains
+
+In additional to configuring the build with rules, bazel provides a additional
+primitive, the toolchain. Toolchains provide default compilers and arguments
+for those compilers. For the native c++ rules, this is relied upon to configure
+the many flags required for cross compilation.
+
+For most iOS projects, configuring the c++ toolchain isn't required. Defining a
+custom toolchain is useful to gives full control over compiler invocations or
+to customize the compiler being invoked.  Please see the [toolchain
+documentation](https://docs.bazel.build/versions/master/tutorial/cc-toolchain-config.html)
+to learn more.
+
+### Generated Xcode projects
+
+In an Xcode world, folks checked in a project which contains an listing of
+files, build settings, and IDE state. When the project scales, the project file
+model breaks down. Auditing and code reviewing config changes in Xcode projects
+quickly becomes difficult. With Bazel, human readable `BUILD` files are the
+source of truth for build and Xcode configuration. Tools like
+[XCHammer](https://github.com/pinterest/xchammer), and
+[Tulsi](https://github.com/bazelbuild/tulsi) use an aspect to traverse the
+build graph and extract metadata required to generate a project.  These tools
+make it easier to manage the project and generate on demand - not needing to
+check it in.
+
+XCHammer provides a rule to [bazel build Xcode
+projects](ihttps://github.com/pinterest/xchammer#bazel-build-xcode-projects).
+Simply declare the rule with the project and Bazel build the target.
+```
+load("@xchammer_resources//:xcodeproject.bzl", "xcode_project")
+xcode_project(
+    name = "MyProject",
+    targets = [ "//ios-app:ios-app" ],
+    paths = [ "**" ],
+)
+```
+
+Like any other Bazel target, it's built from the command line
+```
+bazel build :MyProject
+```
+
+The rule definition for the `xcode_project` may be found in the [github
+repository](https://github.com/pinterest/xchammer/blob/master/BazelExtensions/xcodeproject.bzl).
+Simply put, the aspect traverses sources, and invokes the `xchammer` binary with
+a JSON file. Internally, XCHammer instantiates
+[`XcodeGen`](https://github.com/yonaskolb/XcodeGen) types and writes them out to
+disk with `xcodeproj`.
+
+The workflow for generating an Xcode project with XCHammer is quite simple:
+```
+command line -> bazel -> rule -> xchammer -> xcodegen -> xcodeproj
+```
+
+Generators like XCHammer and Tulsi take care of integrating Bazel into the IDE.
+Bazel builds are invoked a shell script build phase from the IDE. Basically,
+Xcode shells out to Bazel to produce the application, and then Xcode picks up
+the product from the derived data path. Performing a Bazel build from Xcode.
+```
+play button -> shell script build phase -> bazel build ios-app
+```
+
```

Comment:
This was a good explainer, it might be good to diagram of what was described above to help user form a mental model, explain this to their team members, and have a quick reference

Reply:
Good suggestion, I'll add a diagram of this phase

Reply:
It's really useful and enlightening to have the clarification that the IDE calls bazel via a shell script phase, and then Xcode IDE automatically picks up on the results of bazel building from DerivedData.

Reply:
Cool glad it made sense. We're planning to build out better xchammer documentation to describe how it works in more depth, this document touches on the higher level concepts that people have used to build Xcode generators with Bazel. I think @kastiglione might have also wrote on this.

Reply:
I did discuss this in our bazelcon talk.

Reply:
Ahh much appreciated @kastiglione I remember this and great talk BTW. I'll link this up and a few other Bazel con talks to the doc.